File: Day13/ej2.py

laser_security = []
laser_security_end = []
max_num = 0

laser_security = open("input.txt", "r").read().split("\n")

for line in range(len(laser_security)):
  laser_security[line] = laser_security[line].split(": ")    
  laser_security[line] = [int(laser_security[line][0]),[False]+[True]*(int(laser_security[line][1])-1)]
  max_num = int(laser_security[line][0])

laser_security_end = [[[True],0]]*(max_num+1)

for line in laser_security:
  laser_security_end[line[0]] = [list(line[1]),0]

laser_security = list(laser_security_end)

respuesta = [[]]*len(laser_security)

index_mayor = 0
for play in range(2000000000):
  print(play)
  resi = True
  for res in range(len(respuesta)):
    respuesta[res] = respuesta[res] + [laser_security[res][0][0]]
  if len(respuesta[0]) > len(laser_security):
    for i in range(len(respuesta)):
      respuesta[i].pop(0)
  for line in range(len(laser_security)): 
    if laser_security[line][0][0] == False:
      laser_security[line][1] = 0  
    if laser_security[line][0][-1] == False:
      laser_security[line][1] = 1 
    if laser_security[line][1] == 0:
      laser_security[line][0].insert(0,laser_security[line][0].pop())
    else:
      laser_security[line][0].append(laser_security[line][0].pop(0))
  if(len(respuesta) == len(respuesta[len(respuesta)-1])):
    for i in range(len(respuesta)):
      resi = resi & respuesta[i][i]
    if resi:
      print("I have the solution: "+str(1+index_mayor-len(laser_security)))
      break
  index_mayor += 1

#for res in respuesta:
  #print(res)

File: Day13/ej2.rb
laser_security = []
score = 0
max = 0
File.open("input.txt", "r") do |f|
  f.each_line do |line|
    aux_array = line.split(": ")  
    max = aux_array[0].to_i     
    laser_security.push([aux_array[0].to_i,Array.new((aux_array[1].to_i-1), true).unshift(false),0])
  end
end 

state = []
state.push(laser_security)

state_index = 0
for jj in 0..2000000000
  puts ("vuelta "+jj.to_s)
  respuesta = true  
  state_index += 1
  dex = 0
  laser_security = state[jj].dup
  for i in 0..max
    if dex < 0
      dex = dex + 1
    else
      if i == laser_security[dex][0]
        #puts laser_security[dex][1][0]
        respuesta = respuesta & laser_security[dex][1][0]
        if laser_security[dex][1][0] == false
          score = score + i*laser_security[dex][1].length
        end
        dex = dex + 1
      end    
    end

    for laser in laser_security

      if laser[1].first == false
        laser[2] = 0  
      end
      if laser[1].last == false
        laser[2] = 1 
      end
      if laser[2] == 0
        laser[1].unshift(laser[1].pop())
      else
        laser[1].push(laser[1].shift())
      end
    end  
    if state_index + i >= state.length      
      state.push(laser_security.dup)
    end  
  end  
  if respuesta
    puts ("vuelta "+jj.to_s)
    break
  end
end
